index.js

```diff
@@ -18,149 +10,197 @@ var mkdirp = require('mkdirp');
  * function is not passed.
  *
  * ```js
- * var writeFile = require('write');
- * writeFile('foo.txt', 'This is content...', function(err) {
- *   if (err) console.log(err);
- * });
+ * const write = require('write');
+ *
+ * // async/await
+ * (async () => {
+ *   await write('foo.txt', 'This is content...');
+ * })();
  *
  * // promise
- * writeFile('foo.txt', 'This is content...')
- *   .then(function() {
+ * write('foo.txt', 'This is content...')
+ *   .then(() => {
  *     // do stuff
  *   });
+ *
+ * // callback
+ * write('foo.txt', 'This is content...', err => {
+ *   // do stuff with err
+ * });
  * ```
- * @name writeFile
- * @param {string|Buffer|integer} `filepath` filepath or file descriptor.
- * @param {string|Buffer|Uint8Array} `data` String to write to disk.
- * @param {object} `options` Options to pass to [fs.writeFile][fs]{#fs_fs_writefile_file_data_options_callback} and/or [mkdirp][]
+ * @name write
+ * @param {String} `filepath` file path.
+ * @param {String|Buffer|Uint8Array} `data` Data to write.
+ * @param {Object} `options` Options to pass to [fs.writeFile][writefile]
  * @param {Function} `callback` (optional) If no callback is provided, a promise is returned.
  * @api public
  */
 
-function writeFile(filepath, data, options, cb) {
+const write = (filepath, data, options, callback) => {
   if (typeof options === 'function') {
-    cb = options;
+    callback = options;
     options = {};
   }
 
-  if (typeof cb !== 'function') {
-    return writeFile.promise.apply(null, arguments);
-  }
+  const opts = { encoding: 'utf8', ...options };
+  const destpath = opts.increment ? incrementName(filepath) : filepath;
```

Comment:
Great idea! I like that you're doing the check here and incrementing here. It makes it simple later on to just do a file exists check.